esp-wifi/src/lib.rs

```diff
@@ -230,35 +230,6 @@ pub fn initialize(
     radio_clocks: hal::peripherals::RADIO_CLK,
     clocks: &Clocks,
 ) -> Result<EspWifiInitialization, InitializationError> {
-    #[cfg(any(esp32, esp32s3, esp32s2))]
-    const MAX_CLOCK: u32 = 240;
-
-    #[cfg(any(esp32c3, esp32c6))]
-    const MAX_CLOCK: u32 = 160;
-
-    #[cfg(esp32c2)]
-    const MAX_CLOCK: u32 = 120;
-
-    #[cfg(esp32h2)]
-    const MAX_CLOCK: u32 = 96;
-
-    if clocks.cpu_clock != MegahertzU32::MHz(MAX_CLOCK) {
-        return Err(InitializationError::WrongClockConfig);
-    }
-
-    #[cfg(esp32s3)]
-    unsafe {
-        // should be done by the HAL in `ClockControl::configure`
```

Comment:
@bjoernQ I guess we still need this, and we should take the time now to put it in esp-hal's clock module?

Reply:
I'm not entirely sure how this came to be in the first place. I tried tracing the origin of this code and I only found this place https://github.com/gemesa/esp32-phantom/blob/main/esp-wifi-promiscuous.patch

Reply:
@bjoernQ does something like that look reasonable to you?

```patch
diff --git a/esp-hal/src/clock/clocks_ll/esp32s3.rs b/esp-hal/src/clock/clocks_ll/esp32s3.rs
index 83e2afc8..9adf2859 100644
--- a/esp-hal/src/clock/clocks_ll/esp32s3.rs
+++ b/esp-hal/src/clock/clocks_ll/esp32s3.rs
@@ -1,4 +1,9 @@
-use crate::clock::CpuClock;
+use core::ops::Div;
+
+use crate::{
+    clock::{Clock, CpuClock},
+    rom,
+};
 
 pub(crate) fn set_cpu_clock(cpu_clock_speed: CpuClock) {
     let system_control = unsafe { &*crate::peripherals::SYSTEM::PTR };
@@ -19,3 +24,8 @@ pub(crate) fn set_cpu_clock(cpu_clock_speed: CpuClock) {
         });
     }
 }
+
+pub(crate) fn ets_update_cpu_frequency_rom(cpu_clock_speed: CpuClock) {
+    let ticks_per_us = cpu_clock_speed.frequency().div(1000000);
+    rom::ets_update_cpu_frequency_rom(ticks_per_us.raw());
+}
diff --git a/esp-hal/src/clock/mod.rs b/esp-hal/src/clock/mod.rs
index 97f46647..700f9d1f 100644
--- a/esp-hal/src/clock/mod.rs
+++ b/esp-hal/src/clock/mod.rs
@@ -665,15 +665,7 @@ impl<'d> ClockControl<'d> {
     pub fn boot_defaults(
         clock_control: impl Peripheral<P = SystemClockControl> + 'd,
     ) -> ClockControl<'d> {
-        ClockControl {
-            _private: clock_control.into_ref(),
-            desired_rates: RawClocks {
-                cpu_clock: HertzU32::MHz(80),
-                apb_clock: HertzU32::MHz(80),
-                xtal_clock: HertzU32::MHz(40),
-                crypto_pwm_clock: HertzU32::MHz(160),
-            },
-        }
+        Self::configure(clock_control, CpuClock::Clock80MHz)
     }
 
     /// Configure the CPU clock speed.
@@ -682,6 +674,7 @@ impl<'d> ClockControl<'d> {
         cpu_clock_speed: CpuClock,
     ) -> ClockControl<'d> {
         clocks_ll::set_cpu_clock(cpu_clock_speed);
+        clocks_ll::ets_update_cpu_frequency_rom(cpu_clock_speed);
 
         ClockControl {
             _private: clock_control.into_ref(),
```

Also note that if the back and forth is inconvenient, I'm equally happy you guys doing the patch, it seemed a fun challenge to contribute it myself, but it's obvious I have more questions than solutions.

Reply:
No problem! Maybe the function should be named differently since we already use the same name in the `rom` module.

We probably should not change `boot_defaults` to actually configure the clocks - even the call to `ets_update_cpu_frequency_rom` shouldn't be needed for boot_defaults (but is needed whenever the clocks are actually changed)

Reply:
I've decided to merge it with `set_cpu_clock` since it seems closely related to what `rom::ets_update_cpu_frequency_rom` does.